stackslib/src/net/httpcore.rs

```diff
@@ -0,0 +1,1625 @@
+    /// already sent.  It gets called from ProtocolFamily::read_preamble().
+    ///
+    /// This method will set up this state machine to consume the message associated with this
+    /// premable, if the response is chunked.
+    fn set_preamble(&mut self, preamble: &StacksHttpPreamble) -> Result<(), NetError> {
+        match preamble {
+            StacksHttpPreamble::Response(ref http_response_preamble) => {
+                // we can only receive a response if we're expecting it
+                if self.request_handler_index.is_none() {
+                    return Err(NetError::DeserializeError(
+                        "Unexpected HTTP response: no active request handler".to_string(),
+                    ));
+                }
+                if http_response_preamble.is_chunked() {
+                    // we can only receive one response at a time
+                    if self.reply.is_some() {
+                        test_debug!("Have pending reply already");
+                        return Err(NetError::InProgress);
+                    }
+
+                    self.set_pending(http_response_preamble);
+                }
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+
+    /// Clear any pending response state -- i.e. due to a failed request.
+    fn reset(&mut self) -> () {
+        self.request_handler_index = None;
+        self.reply = None;
+    }
+
+    /// Used for processing chunk-encoded streams.
+    /// Given the preamble and a Read, stream the bytes into a chunk-decoder.  Return the decoded
+    /// bytes if we decode an entire stream.  Always return the number of bytes consumed.
+    /// Returns Ok((Some(decoded bytes we got, total number of encoded bytes), number of bytes gotten in this call)) if we're done decoding.
+    /// Returns Ok((None, number of bytes gotten in this call)) if there's more to decode.
+    pub fn consume_data<R: Read>(
+        &mut self,
+        preamble: &HttpResponsePreamble,
+        fd: &mut R,
+    ) -> Result<(Option<(Vec<u8>, usize)>, usize), NetError> {
+        assert!(preamble.is_chunked());
+        assert!(self.reply.is_some());
+
+        if let Some(reply) = self.reply.as_mut() {
+            match reply.stream.consume_data(fd) {
+                Ok(res) => match res {
+                    (None, sz) => Ok((None, sz)),
+                    (Some((byte_vec, bytes_total)), sz) => {
+                        // done receiving
+                        self.reply = None;
+                        Ok((Some((byte_vec, bytes_total)), sz))
+                    }
+                },
+                Err(e) => {
+                    // broken stream
+                    self.reset();
+                    Err(e)
+                }
+            }
+        } else {
+            unreachable!();
+        }
+    }
+
+    /// Calculate the search window for \r\n\r\n
+    fn body_start_search_window(&self, i: usize, buf: &[u8]) -> [u8; 4] {
```

Comment:
Not sure this is better as its also confusing to me but you could simplify this code to 

```rust
fn body_start_search_window(&self, i: usize, buf: &[u8]) -> [u8; 4] {
    let mut window = [0; 4];
    for j in 0..4 {
        window[j] = if i < 4 - j {
            self.last_four_preamble_bytes[j]
        } else {
            buf[i - 4 + j]
        };
    }
    window
}
```

Reply:
Sure, will add comments.